extractor/src/main/java/org/schabi/newpipe/extractor/services/youtube/YoutubeDashManifestCreator.java

```diff
@@ -0,0 +1,1887 @@
+     *                              generates manifests of post DVR livestreams
+     * @throws YoutubeDashManifestCreationException if something goes wrong when generating or
+     *                                              appending the {@code <S>} element to the
+     *                                              document
+     */
+    private static void generateSegmentElementForPostLiveDvrStreams(
+            @Nonnull final Document document,
+            final int targetDurationSeconds,
+            @Nonnull final String segmentCount) throws YoutubeDashManifestCreationException {
+        try {
+            final Element segmentTimelineElement = (Element) document.getElementsByTagName(
+                    "SegmentTimeline").item(0);
+            final Element sElement = document.createElement("S");
+
+            final Attr dAttribute = document.createAttribute("d");
+            dAttribute.setValue(String.valueOf(targetDurationSeconds * 1000));
+            sElement.setAttributeNode(dAttribute);
+
+            final Attr rAttribute = document.createAttribute("r");
+            rAttribute.setValue(segmentCount);
+            sElement.setAttributeNode(rAttribute);
+
+            segmentTimelineElement.appendChild(sElement);
+        } catch (final DOMException e) {
+            throw new YoutubeDashManifestCreationException(
+                    "Could not generate or append the segment (S) elements of the DASH manifest to the document", e);
+        }
+    }
+
+    /**
+     * Convert a DASH manifest {@link Document document} to a string.
+     *
+     * @param originalBaseStreamingUrl the original base URL of the stream
+     * @param document                 the document to be converted
+     * @param manifestCreatorCache     the {@link ManifestCreatorCache} on which store the
+     *                                 string generated (which is either
+     *                                 {@link #GENERATED_OTF_MANIFESTS},
+     *                                 {@link #GENERATED_POST_LIVE_DVR_STREAMS_MANIFESTS} or
+     *                                 {@link #GENERATED_PROGRESSIVE_STREAMS_MANIFESTS})
+     * @return the DASH manifest {@link Document document} converted to a string
+     * @throws YoutubeDashManifestCreationException if something goes wrong when converting the
+     *                                              {@link Document document}
+     */
+    private static String buildResult(
+            @Nonnull final String originalBaseStreamingUrl,
+            @Nonnull final Document document,
+            @Nonnull final ManifestCreatorCache<String, String> manifestCreatorCache)
+            throws YoutubeDashManifestCreationException {
+        try {
+            final StringWriter result = new StringWriter();
+            final TransformerFactory transformerFactory = TransformerFactory.newInstance();
+
+            final Transformer transformer = transformerFactory.newTransformer();
```

Comment:
Yes, I know, but how to restrict this without unauthorizing access and use of transformers by extractor clients? Do you have an idea?

Reply:
@TiA4f8R 
Not sure what you mean exactly but https://rules.sonarsource.com/java/RSPEC-2755 states the following:
```java
TransformerFactory factory = javax.xml.transform.TransformerFactory.newInstance();
// to be compliant, prohibit the use of all protocols by external entities:
factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
factory.setAttribute(XMLConstants.ACCESS_EXTERNAL_STYLESHEET, "");
```

Reply:
I tried this a while, and it disallows everything outside the extractor to build manifests of post live manifests (an exception is thrown).

So do you have another idea?

Reply:
I am not sure I understand the problem here. Isn't the warning related to the XML transformer that might try to load data from network? Can't that just be disabled, since we don't need schema validation or anything like that and all data is already in the XML?

> I tried this a while, and it disallows everything outside the extractor to build manifests of post live manifests (an exception is thrown).

What? Which exception? Why would disabling loading XML entities from network disallow calling this function from outside the extractor? I do not understand...

Reply:
I just cannot create DASH manifests at all (an exception like a `SecurityException` is thrown).

Reply:
> I am not sure I understand the problem here.

10sec googling returns at the first place: https://owasp.org/www-community/vulnerabilities/XML_External_Entity_(XXE)_Processing


> I just cannot create DASH manifests at all (an exception like a SecurityException is thrown).

Full stacktrace pls

Reply:
```
org.schabi.newpipe.extractor.services.youtube.YoutubeDashManifestCreator$YoutubeDashManifestCreationException: Could not convert the DASH manifest generated to a string
        at org.schabi.newpipe.extractor.services.youtube.YoutubeDashManifestCreator.buildResult(YoutubeDashManifestCreator.java:1619)
        at org.schabi.newpipe.extractor.services.youtube.YoutubeDashManifestCreator.createDashManifestFromOtfStreamingUrl(YoutubeDashManifestCreator.java:294)
        at org.schabi.newpipe.player.resolver.PlaybackResolver.generateDashManifestOfYoutubeDashStream(PlaybackResolver.java:188)
        at org.schabi.newpipe.player.resolver.PlaybackResolver.buildDashMediaSource(PlaybackResolver.java:158)
        at org.schabi.newpipe.player.resolver.PlaybackResolver.buildMediaSource(PlaybackResolver.java:104)
        at org.schabi.newpipe.player.resolver.VideoPlaybackResolver.resolve(VideoPlaybackResolver.java:92)
        at org.schabi.newpipe.player.Player.sourceOf(Player.java:3352)
        at org.schabi.newpipe.player.playback.MediaSourceManager.lambda$getLoadedMediaSource$3$org-schabi-newpipe-player-playback-MediaSourceManager(MediaSourceManager.java:419)
        at org.schabi.newpipe.player.playback.MediaSourceManager$$ExternalSyntheticLambda2.apply(Unknown Source:6)
        at io.reactivex.rxjava3.internal.operators.single.SingleMap$MapSingleObserver.onSuccess(SingleMap.java:58)
        at io.reactivex.rxjava3.internal.operators.single.SingleDoOnError$DoOnError.onSuccess(SingleDoOnError.java:52)
        at io.reactivex.rxjava3.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.onSuccess(SingleSubscribeOn.java:68)
        at io.reactivex.rxjava3.internal.operators.maybe.MaybeToSingle$ToSingleMaybeSubscriber.onSuccess(MaybeToSingle.java:83)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableElementAtMaybe$ElementAtSubscriber.onNext(FlowableElementAtMaybe.java:80)
        at io.reactivex.rxjava3.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.drain(MaybeConcatArray.java:136)
        at io.reactivex.rxjava3.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.request(MaybeConcatArray.java:78)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableElementAtMaybe$ElementAtSubscriber.onSubscribe(FlowableElementAtMaybe.java:66)
        at io.reactivex.rxjava3.internal.operators.maybe.MaybeConcatArray.subscribeActual(MaybeConcatArray.java:42)
        at io.reactivex.rxjava3.core.Flowable.subscribe(Flowable.java:15868)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableElementAtMaybe.subscribeActual(FlowableElementAtMaybe.java:36)
        at io.reactivex.rxjava3.core.Maybe.subscribe(Maybe.java:5330)
        at io.reactivex.rxjava3.internal.operators.maybe.MaybeToSingle.subscribeActual(MaybeToSingle.java:46)
        at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4813)
        at io.reactivex.rxjava3.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89)
        at io.reactivex.rxjava3.core.Scheduler$DisposeTask.run(Scheduler.java:644)
        at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
        at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:929)
     Caused by: java.lang.IllegalArgumentException: Not supported: http://javax.xml.XMLConstants/property/accessExternalDTD
        at org.apache.xalan.processor.TransformerFactoryImpl.setAttribute(TransformerFactoryImpl.java:580)
        at org.schabi.newpipe.extractor.services.youtube.YoutubeDashManifestCreator.buildResult(YoutubeDashManifestCreator.java:1607)
        at org.schabi.newpipe.extractor.services.youtube.YoutubeDashManifestCreator.createDashManifestFromOtfStreamingUrl(YoutubeDashManifestCreator.java:294) 
        at org.schabi.newpipe.player.resolver.PlaybackResolver.generateDashManifestOfYoutubeDashStream(PlaybackResolver.java:188) 
        at org.schabi.newpipe.player.resolver.PlaybackResolver.buildDashMediaSource(PlaybackResolver.java:158) 
        at org.schabi.newpipe.player.resolver.PlaybackResolver.buildMediaSource(PlaybackResolver.java:104) 
        at org.schabi.newpipe.player.resolver.VideoPlaybackResolver.resolve(VideoPlaybackResolver.java:92) 
        at org.schabi.newpipe.player.Player.sourceOf(Player.java:3352) 
        at org.schabi.newpipe.player.playback.MediaSourceManager.lambda$getLoadedMediaSource$3$org-schabi-newpipe-player-playback-MediaSourceManager(MediaSourceManager.java:419) 
        at org.schabi.newpipe.player.playback.MediaSourceManager$$ExternalSyntheticLambda2.apply(Unknown Source:6) 
        at io.reactivex.rxjava3.internal.operators.single.SingleMap$MapSingleObserver.onSuccess(SingleMap.java:58) 
        at io.reactivex.rxjava3.internal.operators.single.SingleDoOnError$DoOnError.onSuccess(SingleDoOnError.java:52) 
        at io.reactivex.rxjava3.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.onSuccess(SingleSubscribeOn.java:68) 
        at io.reactivex.rxjava3.internal.operators.maybe.MaybeToSingle$ToSingleMaybeSubscriber.onSuccess(MaybeToSingle.java:83) 
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableElementAtMaybe$ElementAtSubscriber.onNext(FlowableElementAtMaybe.java:80) 
        at io.reactivex.rxjava3.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.drain(MaybeConcatArray.java:136) 
        at io.reactivex.rxjava3.internal.operators.maybe.MaybeConcatArray$ConcatMaybeObserver.request(MaybeConcatArray.java:78) 
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableElementAtMaybe$ElementAtSubscriber.onSubscribe(FlowableElementAtMaybe.java:66) 
        at io.reactivex.rxjava3.internal.operators.maybe.MaybeConcatArray.subscribeActual(MaybeConcatArray.java:42) 
        at io.reactivex.rxjava3.core.Flowable.subscribe(Flowable.java:15868) 
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableElementAtMaybe.subscribeActual(FlowableElementAtMaybe.java:36) 
        at io.reactivex.rxjava3.core.Maybe.subscribe(Maybe.java:5330) 
        at io.reactivex.rxjava3.internal.operators.maybe.MaybeToSingle.subscribeActual(MaybeToSingle.java:46) 
        at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4813) 
        at io.reactivex.rxjava3.internal.operators.single.SingleSubscribeOn$SubscribeOnObserver.run(SingleSubscribeOn.java:89) 
        at io.reactivex.rxjava3.core.Scheduler$DisposeTask.run(Scheduler.java:644) 
        at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65) 
        at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56) 
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) 
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:929) 
```

Reply:
What we should do then @litetex?

Reply:
@litetex The issue should have been solved by Stypox. Let us know in another comment/review if it is good for you.